### DISCUSSION TITLE
Improving the auth workflow of Supabase v2

### DISCUSSION THREAD

**Discussion by Olyno** [supabase/supabase#8451] [General]
Hi 👋🏻 
After having experienced the authentication system of version 2, I would like to give feedback and discuss potential improvements to it.

First, congratulations to the team and to the different contributors, because clear improvements have been made in the SDK. It is easier to use, clearer, and this can't be possible without you!

During my test, I could notice that there were several types for the method ``signInWithOtp``:
 - signup
 - invite
 - magiclink
 - recovery
 - email_change
 - sms
 - phone_change

I think that many of the above types are not useful. Let's take a concrete example:
 - A user arrives for the first time on mywebsite.com
 - He wants to create an account
 - He goes to the registration/login page
 - He registers in passwordless.
 - He enters the code he received by email.
 - He is connected

In the above case, we have to use the ``signup`` type to register the user, which makes sense... or maybe not. Let's take a 2nd case:
 - A user arrives for the nth time on mywebsite.com
 - He wants to login to his account
 - He goes to the registration/login page
 - He logs in passwordless.
 - He enters the code he received by email.
 - He is connected

In the 2 cases above, we have the same path for the same result. However, in the 2nd case, we will have to use the ``magiclink`` type, while it can very well be a code.

I think the terms could be improved. Here is my new type proposal:
```diff
- - signup
 - invite
- - magiclink
+ - email
 - recovery
 - email_change
 - sms
 - phone_change
```

Here is a demonstration of this logic in code, with NextJS:

```ts
const verifyOtp = async (code: string, email: string) => {
  if (!email) return;
  const { error: signup_error } = await supabaseClient.auth.verifyOtp({
    email,
    token: code.trim(),
    type: 'signup',
  });
  if (signup_error) {
    const { error } = await supabaseClient.auth.verifyOtp({
      email,
      token: code.trim(),
      type: 'magiclink',
    });
    if (!error) {
      router.push('/');
    } else {
      setError("We couldn't verify your code...");
    }
  } else {
    router.push('/');
  }
};
```

By doing this, we reduce the complexity of the workflow to have the same impact at the end. We simplify the terms so that all possible solutions are in one place. Instead of having two separate types for registering and logging in, having one type named ``email`` automatically registers the user if it doesn't already exist, and then sends them the code to log in. It also allows to stay consistent with its mobile version.

What do you think?

**Comment by landsman** [comment]
yes please, current flow is overkill

**Comment by landsman** [comment]
https://github.com/supabase/gotrue/issues/905#issuecomment-1533128779
